label-application-rule-config-service-api/src/main/proto/org/hypertrace/label/application/rule/config/service/v1/label_application_rule.proto

```diff
@@ -0,0 +1,98 @@
+syntax = "proto3";
+
+package org.hypertrace.label.application.rule.config.service.v1;
+
+option java_multiple_files = true;
+
+message LabelApplicationRule {
+  string id = 1;
+  LabelApplicationRule data = 2;
+}
+
+message CreateLabelApplicationRule {
+  LabelApplicationRule data = 2;
+}
+
+message LabelApplicationRuleData {
+  string name = 1;
+  string description = 2;
+  MatchingCriteria matching_criteria = 3;
+  ApplicationCriteria application_criteria = 4;
+
+  message MatchingCriteria {
+    Type type = 1;
+    repeated Condition conditions = 2;
+
+    enum Type {
+      TYPE_UNSPECIFIED = 0;
+      TYPE_SPAN = 1;
+      TYPE_TRACE = 2;
+    }
+
+    message Condition {
+      string name = 1;
+      Scope scope = 2;
+      string key = 3;
+      Operator operator = 4;
+      optional string value = 5;
+
+      enum Scope {
+        SCOPE_UNSPECIFIED = 0;
+        SCOPE_ATTRIBUTE = 1;
+        SCOPE_REQUEST = 2;
+        SCOPE_RESPONSE = 3;
+      }
+
+      enum Operator {
+        OPERATOR_UNSPECIFIED = 0;
+        OPERATOR_EXISTS = 1;
+        OPERATOR_EQUALS = 2;
+        OPERATOR_CONTAINS = 3;
+        OPERATOR_STARTS_WITH = 4;
+        OPERATOR_ENDS_WITH = 5;
+        OPERATOR_MATCHES_REGEX = 6;
+      }
+    }
+  }
+
+  message ApplicationCriteria {
+    Type type = 1;
+    Scope scope = 2;
+    Action action = 3;
+
+    enum Type {
+      TYPE_UNSPECIFIED = 0;
+      TYPE_ENTITY = 1;
+    }
+
+    enum Scope {
```

Comment:
Can we abstract this so it's not an enum? That might just mean replacing it with a string. The actual entities support in the system are config driven, and should remain so here.

Reply:
okay. However we will have validation around it to ensure that the entity that is being labelled is either API, Service or Domain. Correct?

Reply:
Ideally, no - at least not in code. A runtime check on the other hand could be feasible by a call to entity type system (but config service shouldn't be calling there) to ensure an entity type exists. Not every system will have every entity so we don't want to bake in any assumption if we can help it - either in api or impl.

Reply:
Runtime check is a bit too late. We need to notify user if he/she has selected a wrong scope?

Reply:
By runtime, I mean during validation rather than execution. The distinction I was making with runtime was looking up the allowed scope strings vs defining them in code.

Reply:
Decision here is that validation will be based on supported entity types provided in config in short term.

In the long term, entity type service should provide details about which entity type can be labelled. Once this information is available, it can be fetched at any place for validation purpose. However config service will not pull this detail for validation from entity type service, as it may bring cyclic dependency. The validation will happen at higher layers like UI, GraphQL.